auth_brute_force/models/res_users.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2017 Tecnativa - Jairo Llopis
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+from contextlib import contextmanager
+
+from odoo import SUPERUSER_ID, api, models
+from odoo.exceptions import AccessDenied
+from odoo.http import request
+
+# from threading import current_thread
+
+
+_logger = logging.getLogger(__name__)
+
+
+class ResUsers(models.Model):
+    _inherit = "res.users"
+
+    # HACK https://github.com/odoo/odoo/issues/24183
+    # TODO Remove in v12, and use normal odoo.http.request to get details
+    # @api.model_cr
+    # def _register_hook(self):
+    #     """🐒-patch XML-RPC controller to know remote address."""
+    #     original_fn = wsgi_server.application_unproxied
+    #
+    #     def _patch(environ, start_response):
+    #         current_thread().environ = environ
+    #         return original_fn(environ, start_response)
+    #
+    #     wsgi_server.application_unproxied = _patch
+
+    # Helpers to track authentication attempts
+    @classmethod
+    @contextmanager
+    def _auth_attempt(cls, login):
+        """Start an authentication attempt and track its state."""
+        cls.environ = request.httprequest.environ
```

Comment:
You're using the `request` object from within a model.

When you're writing a controller, you can expect that there'll always be a bound request. But models can be called without a request. For example, during normal unit tests written within a `TransactionCase` subclass. Since there's no request, the test fails.

Since it makes no sense to prevent brute-force attacks when there's no request, you should capture that unbound request error and just pass in case it's produced.

Reply:
@yajo 

Thank you for your pointer. We have fixed this in our recent push(es)